### PR TITLE
[RAPTOR-3570] add target type parameter

### PR DESCRIPTION
**Pull request by yakov-g**
# This repository is public. Do not put here any private DataRobot or customer's data: code, datasets, model artifacts, .etc.

## Summary
Add **mandatory** _target-type_ parameter in inference modes.
Not adding explicit _target-type_ parameter in fit mode, but still resolve it magically. (@4dahalibut and I think it is right for now)

Update all the scritps/tests public envs.
_Forgot to update some READMEs, but you are welcome to review_

There are four commits, which may have sense to review separately:
1. Add target  type param and plumb it everywhere
2. Move reading meta file to the very beginning of drum flow
3. Because of #2, i had to split sklearn training model into reg and bin, so they have proper meta files
4. bump version, update env

## Rationale
Currently we magically detect if target type is binary or regression.
As we are moving towards more target types, we need explicitly have such parameter in drum.

### CODE REVIEW

File: custom_model_runner/datarobot_drum/drum/drum.py [hunk 1]
```diff
@@ -54,6 +55,22 @@ def __init__(self, runtime):
         self.verbose = runtime.options.verbose
         self.run_mode = RunMode(runtime.options.subparser_name)
         self.raw_arguments = sys.argv
+        self.target_type = None
+
+        if self.run_mode != RunMode.NEW:
+            # if target type and class labels are not provided assume it is regression
+            if self.options.target_type is None:
+                if self.options.negative_class_label is None:
+                    self.options.target_type = TargetType.REGRESSION.value
```
Comment (liororama) [inline custom_model_runner/datarobot_drum/drum/drum.py:64]: It would be good to add a log.info here since this is going to be "not the standard behavior" soon

Comment (liororama) [reply custom_model_runner/datarobot_drum/drum/drum.py:64]: Also maybe adding some sort of a deprecation message that say that providing an explicit target type might be required in the future...

File: custom_model_runner/datarobot_drum/drum/drum.py [hunk 2]
```diff
@@ -54,6 +55,22 @@ def __init__(self, runtime):
...
+                elif None not in [
+                    self.options.positive_class_label,
+                    self.options.negative_class_label,
+                ]:
+                    self.options.target_type = TargetType.BINARY.value
+                else:
+                    raise DrumCommonException("Target type can not be resolved")
```
Comment (liororama) [inline custom_model_runner/datarobot_drum/drum/drum.py:71]: Make the error more informative.. like saying no postive or negative classes were provided and no --target type ...

File: tests/drum/test_unstructured_mode.py [hunk 3]
```diff
@@ -44,7 +44,7 @@ def test_unstructured_models_batch(
 
         output = tmp_path / "output"
 
-        cmd = "{} score --code-dir {} --input {} --output {} --unstructured".format(
+        cmd = "{} score --code-dir {} --input {} --output {} --target-type unstructured".format(
```
Comment (liororama) [inline tests/drum/test_unstructured_mode.py:47]: Not specific for this line, but there are not tests for the case we need to guess the target type and fail + no other tests with --target-type?

File: custom_model_runner/datarobot_drum/drum/drum.py (hunk 2 above)
Comment (yakov-g) [reply custom_model_runner/datarobot_drum/drum/drum.py:89]: Eventually, I remove this exception as it doesn't make sense: currently, if labels are not provided, it is regression.

I'll check a bit more if it is already better to make --target-type required

File: custom_model_runner/datarobot_drum/drum/drum.py (hunk 1 above)
Comment (yakov-g) [reply custom_model_runner/datarobot_drum/drum/drum.py:82]: I'll check a bit more if it is already better to make --target-type required

File: tests/drum/test_unstructured_mode.py (hunk 3 above)
Comment (yakov-g) [reply tests/drum/test_unstructured_mode.py:47]: I'll review it deeper, as looks like it is better to plumb `target_type` trough all the predictors, by doing:
```
if binary:
   predict binary
elif regression:
   predict regression
```

**Comment by yakov-g** [comment]
jarvis test functional_drop_in

File: custom_model_runner/datarobot_drum/drum/drum.py (hunk 1 above)
Comment (yakov-g) [reply custom_model_runner/datarobot_drum/drum/drum.py]: Made --target-type mandatory for inference

File: tests/drum/test_unstructured_mode.py (hunk 3 above)
Comment (yakov-g) [reply tests/drum/test_unstructured_mode.py:47]: did it

File: custom_model_runner/datarobot_drum/drum/artifact_predictors/torch_predictor.py [hunk 4]
```diff
@@ -72,14 +74,21 @@ def predict(self, data, model, **kwargs):
         )
         with torch.no_grad():
             predictions = model(data).cpu().data.numpy()
-        if self.positive_class_label is not None and self.negative_class_label is not None:
+        print("im in torch")
```
Comment (meltzg) [inline custom_model_runner/datarobot_drum/drum/artifact_predictors/torch_predictor.py:77]: looks like a leftover debug statement

Comment (yakov-g) [reply custom_model_runner/datarobot_drum/drum/artifact_predictors/torch_predictor.py:77]: :)

**Review by meltzg — APPROVED** [review]

**Comment by yakov-g** [comment]
jarvis test functional_drop_in

**Comment by yakov-g** [comment]
jarvis test functional_drop_in

**Comment by yakov-g** [comment]
jarvis test functional_drop_in

**Comment by yakov-g** [comment]
jarvis test functional_drop_in

File: custom_model_runner/datarobot_drum/drum/drum.py [hunk 5]
```diff
@@ -42,18 +45,50 @@
...
 
+from strictyaml import Bool, Int, load, Map, Optional, Str, YAMLError
+
 SERVER_PIPELINE = "prediction_server_pipeline.json.j2"
 PREDICTOR_PIPELINE = "prediction_pipeline.json.j2"
 
 
 class CMRunner:
     def __init__(self, runtime):
+        def _read_metadata(code_dir):
```
Comment (4dahalibut) [inline custom_model_runner/datarobot_drum/drum/drum.py:56]: I'd rather this not be an internal method...

File: custom_model_runner/datarobot_drum/drum/drum.py [hunk 6]
```diff
@@ -42,18 +45,50 @@
...
+        # doesn't clash with the target type from model config file
+        if self.options.model_config is not None:
+            model_config_target_type = TargetType(self.options.model_config["targetType"])
+            if self.target_type is None:
+                self.target_type = TargetType(model_config_target_type)
+            elif self.target_type != model_config_target_type:
+                raise DrumCommonException(
+                    "Resolved target type doesn't match target type from model config file."
+                    "If model config file is used used, make --target-type values match."
+                )
```
Comment (4dahalibut) [inline custom_model_runner/datarobot_drum/drum/drum.py:91]: can you extract this?

File: custom_model_runner/datarobot_drum/drum/drum.py [hunk 7]
```diff
@@ -66,6 +101,10 @@ def _config_logger(options):
         logger.setLevel(LOG_LEVELS[options.logging_level])
         return logger
 
+    def _set_target_type(self, target_type):
+        self.target_type = target_type
+        self.options.target_type = target_type.value
+
```
Comment (4dahalibut) [inline custom_model_runner/datarobot_drum/drum/drum.py:107]: I'm not a fan of this information being stored twice.

File: custom_model_runner/datarobot_drum/drum/language_predictors/r_common_code/common.R [hunk 8]
```diff
@@ -1,4 +1,15 @@
 
+TargetTypeEnum <- function() {
```
Comment (4dahalibut) [inline custom_model_runner/datarobot_drum/drum/language_predictors/r_common_code/common.R:2]: whats this do?

File: custom_model_runner/datarobot_drum/drum/language_predictors/r_predictor/score.R [hunk 9]
```diff
@@ -112,9 +116,11 @@ model_predict <- function(data, model, positive_class_label=NULL, negative_class
             predictions <- predictions[, c(2, 1)]
         }
         names(predictions) <- provided_labels
-    } else {
+    } else if (target_type == TargetType$REGRESSION) {
         predictions <- data.frame(stats::predict(model, data))
         names(predictions) <- c(REGRESSION_PRED_COLUMN_NAME)
+    } else {
```
Comment (4dahalibut) [inline custom_model_runner/datarobot_drum/drum/language_predictors/r_predictor/score.R:122]: what about anomaly?

File: custom_model_runner/datarobot_drum/drum/push.py [hunk 10]
```diff
@@ -3,60 +3,30 @@
...
-    return model_config
+    return options.model_config
 
 
 def _convert_target_type(unconverted_target_type):
-    if unconverted_target_type == "regression":
+    if unconverted_target_type == TargetType.REGRESSION.value:
         return dr_client.TARGET_TYPE.REGRESSION
-    elif unconverted_target_type == "binary":
+    elif unconverted_target_type == TargetType.BINARY.value:
```
Comment (4dahalibut) [inline custom_model_runner/datarobot_drum/drum/push.py:29]: @jlpolit did you add anomaly to drum push? doesn't look like it?

File: custom_model_runner/datarobot_drum/drum/language_predictors/r_predictor/score.R (hunk 9 above)
Comment (yakov-g) [reply custom_model_runner/datarobot_drum/drum/language_predictors/r_predictor/score.R:122]: Yes, forgot it

File: custom_model_runner/datarobot_drum/drum/push.py (hunk 10 above)
Comment (jlpolit) [reply custom_model_runner/datarobot_drum/drum/push.py:29]: I did not - it's going to require some changes to the api client, which I have a PR open for but was told to hold off until they published a version....which reminds me I should really check on that

File: model_templates/training/python3_sklearn_regression/create_pipeline.py [hunk 11]
```diff
@@ -0,0 +1,129 @@
...
+### Modeling
+##############################
+
+
+def make_classifier(X):
+    return Pipeline(
+        steps=[("preprocessing", dense_preprocessing_pipeline), ("model", LogisticRegression())]
+    )
+
+
```
Comment (4dahalibut) [inline model_templates/training/python3_sklearn_regression/create_pipeline.py:127]: please delete this and vice versa in other template.

File: custom_model_runner/datarobot_drum/drum/push.py (hunk 10 above)
Comment (jlpolit) [reply custom_model_runner/datarobot_drum/drum/push.py:29]: (this is that thing https://github.com/datarobot/public_api_client/pull/1307 )

File: tests/drum/test_fit.py [hunk 12]
```diff
@@ -119,6 +118,69 @@ def test_fit(
             cmd, "Failed in {} command line! {}".format(ArgumentsOptions.MAIN_COMMAND, cmd)
         )
 
+    @pytest.mark.parametrize(
+        "framework", [RDS, SKLEARN_BINARY, SKLEARN_REGRESSION, XGB, KERAS, PYTORCH]
+    )
+    @pytest.mark.parametrize("problem", [BINARY_TEXT, REGRESSION, ANOMALY])
+    @pytest.mark.parametrize("docker", [DOCKER_PYTHON_SKLEARN, None])
+    @pytest.mark.parametrize("weights", [WEIGHTS_CSV, WEIGHTS_ARGS, None])
+    def test_fit(
```
Comment (4dahalibut) [inline tests/drum/test_fit.py:127]: ???

File: tests/functional/test_inference_model_templates.py [hunk 13]
```diff
@@ -11,15 +11,15 @@ class TestInferenceModelTemplates(object):
...
-            (
-                "inference/java_codegen",
-                "java",
-                "java_drop_in_env",
-                "regression_testing_data",
-                "MEDV",
-                None,
-                None,
-            ),
+            # (
```
Comment (4dahalibut) [inline tests/functional/test_inference_model_templates.py:14]: ????

**Review by 4dahalibut — CHANGES_REQUESTED** [review]
you commented out the functional tests, as well as other things.

File: tests/functional/test_inference_model_templates.py (hunk 13 above)
Comment (yakov-g) [reply tests/functional/test_inference_model_templates.py:14]: there is tmp patch sneaked in, for testing

File: tests/drum/test_fit.py (hunk 12 above)
Comment (yakov-g) [reply tests/drum/test_fit.py:127]: I added  `test_fit_for_use_output_and_nested` which tests only nesting and output dir with one ML problem, thus i removed these params permutations from `test_fit` which significantly reduces number of test cases.

File: custom_model_runner/datarobot_drum/drum/drum.py (hunk 7 above)
Comment (yakov-g) [reply custom_model_runner/datarobot_drum/drum/drum.py:107]: the best would be to save model related params in a separate structure, like context, instead of passing options everywhere.

File: custom_model_runner/datarobot_drum/drum/drum.py (hunk 5 above)
Comment (yakov-g) [reply custom_model_runner/datarobot_drum/drum/drum.py:56]: moved to common.py

File: model_templates/training/python3_sklearn_regression/create_pipeline.py (hunk 11 above)
Comment (yakov-g) [reply model_templates/training/python3_sklearn_regression/create_pipeline.py:127]: done

File: custom_model_runner/datarobot_drum/drum/drum.py (hunk 7 above)
Comment (yakov-g) [reply custom_model_runner/datarobot_drum/drum/drum.py]: done

File: custom_model_runner/datarobot_drum/drum/language_predictors/r_predictor/score.R (hunk 9 above)
Comment (yakov-g) [reply custom_model_runner/datarobot_drum/drum/language_predictors/r_predictor/score.R]: done

File: custom_model_runner/datarobot_drum/drum/drum.py (hunk 6 above)
Comment (yakov-g) [reply custom_model_runner/datarobot_drum/drum/drum.py:83]: done

**Comment by yakov-g** [comment]
@4dahalibut feel free to approve :)

**Comment by yakov-g** [comment]
jarvis test functional_drop_in

**Review by 4dahalibut — APPROVED** [review]

**Comment by datarobot-jarvis** [comment]
:dollar: Total cost of executing tests in Jenkins: $**22.68** USD.

[What is this all about?](http://bit.ly/2BahK4V)
